compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoPackUnPack.cpp

```diff
@@ -417,39 +403,68 @@ lowerOpWithEncoding(RewriterBase &rewriter, tensor::EmptyOp emptyOp,
+///  - linalg::LinalgOp that `isaContractionOpInterface`
+///  - linalg::FillOp
+///  - element-wise linalg::GenericOp with single input and output
+static FailureOr<Operation *> lowerOpWithEncoding(
+    RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+    ValueRange convertedInputOperands, ValueRange convertedOutputOperands,
+    MaterializeEncodingFn materializeEncodingFn, MaterializeEncodingValueFn) {
+  // ContractionOpInterface
+  if (linalg::isaContractionOpInterface(linalgOp)) {
+    SmallVector<Value> operands;
+    operands.append(convertedInputOperands.begin(),
+                    convertedInputOperands.end());
+    operands.append(convertedOutputOperands.begin(),
+                    convertedOutputOperands.end());
+    return lowerContractionOpWithEncoding(rewriter, linalgOp, operands,
+                                          materializeEncodingFn);
+  }
+
+  // linalg::FillOp
+  if (auto fillOp = dyn_cast<linalg::FillOp>(linalgOp.getOperation())) {
+    if (!fillOp.hasPureTensorSemantics())
+      return failure();
+    Operation *materializedFillOp = rewriter.create<linalg::FillOp>(
+        fillOp.getLoc(), convertedOutputOperands[0].getType(),
+        convertedInputOperands, convertedOutputOperands);
+    return materializedFillOp;
+  }
+
+  // element-wise linalg::GenericOp
+  if (auto genericOp = dyn_cast<linalg::GenericOp>(linalgOp.getOperation())) {
+    if (!genericOp.hasPureTensorSemantics() || !isElementwise(genericOp) ||
+        genericOp.getNumDpsInputs() != 1 || genericOp.getNumDpsInits() != 1) {
+      return rewriter.notifyMatchFailure(genericOp,
+                                         "linalg.generic op is not elementwise "
+                                         "with single input and single output");
+    }
+    if (!llvm::all_of(genericOp.getIndexingMapsArray(),
+                      [](AffineMap m) { return m.isIdentity(); })) {
+      return rewriter.notifyMatchFailure(
+          genericOp, "indexing maps are not all identity maps");
+    }
+    auto convertedResultType =
+        convertedOutputOperands[0].getType().cast<RankedTensorType>();
+    SmallVector<AffineMap> maps(
+        2, AffineMap::getMultiDimIdentityMap(convertedResultType.getRank(),
+                                             rewriter.getContext()));
+    SmallVector<utils::IteratorType> iteratorTypes(
+        convertedResultType.getRank(), utils::IteratorType::parallel);
+    auto materializedGenericOp = rewriter.create<linalg::GenericOp>(
+        genericOp.getLoc(), convertedResultType, convertedInputOperands,
+        convertedOutputOperands, maps, iteratorTypes,
+        /*bodyBuild=*/nullptr, linalg::getPrunedAttributeList(genericOp));
+    rewriter.inlineRegionBefore(genericOp.getRegion(),
+                                materializedGenericOp.getRegion(),
+                                materializedGenericOp.getRegion().begin());
+    return materializedGenericOp.getOperation();
+  }
+
+  return failure();
```

Comment:
I think we can use TypeSwitch for this case. See an example in https://github.com/openxla/iree/blob/eaefd3ff964f14e700b1c468aa2b4f20a4401028/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L2018-L2038